### PR TITLE
Add Kubernetes ip restrictions resource

### PR DESCRIPTION
**Pull request by IamBlueSlime**
Hi!

Here is a Terraform resource for managing Kubernetes cluster's ip restrictions. Also exploded the `types_cloud.go` file into multiple specialized ones.

I tested the new resource on my own cloud project without any issue.

Please note that this is my first time in Go, any tip would be highly welcomed!

### CODE REVIEW

File: ovh/resource_cloud_project_kube_iprestriction.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+func resourceCloudProjectKubeIpRestrictionsCreate(d *schema.ResourceData, meta interface{}) error {
+	return resourceCloudProjectKubeIpRestrictionsUpdate(d, meta)
+}
+
+func resourceCloudProjectKubeIpRestrictionsRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	serviceName := d.Get("service_name").(string)
+	kubeId := d.Get("kube_id").(string)
+
+	endpoint := fmt.Sprintf("/cloud/project/%s/kube/%s/ipRestrictions", serviceName, kubeId)
```
Comment (yanndegat) [inline ovh/resource_cloud_project_kube_iprestriction.go:76]: servicename and kubeid should probably be "url.PathEncoded" when forging endpoints.

File: ovh/resource_cloud_project_kube_iprestrictions_test.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func init() {
```
Comment (yanndegat) [inline ovh/resource_cloud_project_kube_iprestrictions_test.go:15]: this sweeper is probably useless as the ovh_cloud_project_kube existing sweeper will delete all clusters with a name starting the the test prefix. this code can be safely removed

File: ovh/resource_cloud_project_kube_iprestriction.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	return nil
+}
+
+func resourceCloudProjectKubeIpRestrictionsUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	serviceName := d.Get("service_name").(string)
+	kubeId := d.Get("kube_id").(string)
+
+	endpoint := fmt.Sprintf("/cloud/project/%s/kube/%s/ipRestrictions", serviceName, kubeId)
+	params := (&CloudProjectKubeIpRestrictionsCreateOpts{}).FromResource(d)
```
Comment (yanndegat) [inline ovh/resource_cloud_project_kube_iprestriction.go:97]: in types_cloud_project_kube_iprestrictiones, you have 2 types (create & update), 
here we use the CreateOpts type in the Update method., 
the create method points to the the udpdate methode

ihmo, it brings a lot of confustion.

either we have 2 explicit methods, with 2 explicit types
or we can have one method CreateOrUpdate with one type CreateOrUpdateOpts

this way, you can point to this only method in the schema definition.

what do you think ?

**Review by yanndegat — CHANGES_REQUESTED** [review]
Hi @IamBlueSlime

thanks a lot for your great work. i heard that this is a feature waited by some customers. 
so it will help a lot.

i have a few comments which i made inline.

3 overall comments:

- this is great work regarding the go side.
- the doc is missing in the website directory. do you think you can manage to add it ?
  you have several examples.
- could squash 2 of your commits ?
   - Add Kubernetes Iprestrictions type
   - Add Kubernetes Iprestrictions resource



thanks again

File: ovh/resource_cloud_project_kube_iprestriction.go (hunk 3 above)
Comment (IamBlueSlime) [reply ovh/resource_cloud_project_kube_iprestriction.go]: I agree with you on this!
Fixed.

File: ovh/resource_cloud_project_kube_iprestrictions_test.go (hunk 2 above)
Comment (IamBlueSlime) [reply ovh/resource_cloud_project_kube_iprestrictions_test.go]: Fixed.

File: ovh/resource_cloud_project_kube_iprestriction.go (hunk 1 above)
Comment (IamBlueSlime) [reply ovh/resource_cloud_project_kube_iprestriction.go]: Fixed.

File: ovh/resource_cloud_project_kube_iprestriction.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
+package ovh
```
[the remark this answers is not in the document]
Comment (IamBlueSlime) [reply ovh/resource_cloud_project_kube_iprestriction.go]: Fixed.

File: ovh/provider.go [hunk 5]
```diff
@@ -86,6 +86,7 @@ func Provider() *schema.Provider {
 			"ovh_cloud_project_containerregistry_user":                    resourceCloudProjectContainerRegistryUser(),
 			"ovh_cloud_project_kube":                                      resourceCloudProjectKube(),
 			"ovh_cloud_project_kube_nodepool":                             resourceCloudProjectKubeNodePool(),
+			"ovh_cloud_project_kube_iprestrictions":                       resourceCloudProjectKubeIpRestrictions(),
```
[the remark this answers is not in the document]
Comment (IamBlueSlime) [reply ovh/provider.go:88]: Fixed.

**Review by IamBlueSlime — COMMENTED** [review]
Hi @yanndegat, thanks a lot, I've managed to fix all your suggestions including adding the HTML documentation. Don't hesitate if anything is needed.

**Review by yanndegat — APPROVED** [review]

**Comment by yanndegat** [comment]
LGTM thanks

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#361.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#361.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#381.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#381.0 ✘  
* 
  * checks ✘

**Comment by yanndegat** [comment]
hi @IamBlueSlime 

i hit this issue when running your acceptance tests:

```
[INFO] === RUN   TestAccCloudProjectKubeIpRestrictions_basic

[INFO]     provider_test.go:208: Read Cloud Project /cloud/project/9944712e635e41478b28de8534683477 -> status: 'ok', desc: 'Hashicorp Acceptance Tests'

[INFO]     resource_cloud_project_kube_iprestrictions_test.go:39: Step 1/1 error: Check failed: Check 5/5 error: ovh_cloud_project_kube_iprestrictions.iprestrictions: Attribute 'ips' not found

```

**Comment by IamBlueSlime** [comment]
> hi @IamBlueSlime
> 
> i hit this issue when running your acceptance tests:
> 
> ```
> [INFO] === RUN   TestAccCloudProjectKubeIpRestrictions_basic
> 
> [INFO]     provider_test.go:208: Read Cloud Project /cloud/project/9944712e635e41478b28de8534683477 -> status: 'ok', desc: 'Hashicorp Acceptance Tests'
> 
> [INFO]     resource_cloud_project_kube_iprestrictions_test.go:39: Step 1/1 error: Check failed: Check 5/5 error: ovh_cloud_project_kube_iprestrictions.iprestrictions: Attribute 'ips' not found
> ```

Hi, sorry about that. I've read all the code and I can't manage to see where I'm doing things wrong.

The broken test assertion is the following:
```go
					resource.TestCheckResourceAttr(
						"ovh_cloud_project_kube_iprestrictions.iprestrictions", "ips", "[10.42.0.0/16]"),
```
But in fact, this is an attribute required for the resource to be created, and it is correctly exported in the _read_ function, see L85 of `resource_cloud_project_kube_iprestrictions.go`: `d.Set("ips", res)`.

Do you have any tips for me?

File: ovh/resource_cloud_project_kube_iprestrictions_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+					resource.TestCheckResourceAttr(
+						"ovh_cloud_project_kube.cluster", "region", region),
+					resource.TestCheckResourceAttrSet(
+						"ovh_cloud_project_kube.cluster", "kubeconfig"),
+					resource.TestCheckResourceAttr(
+						"ovh_cloud_project_kube.cluster", "name", name),
+					resource.TestCheckResourceAttr(
+						"ovh_cloud_project_kube.cluster", "version", version),
+					resource.TestCheckResourceAttr(
+						"ovh_cloud_project_kube_iprestrictions.iprestrictions", "ips", "[10.42.0.0/16]"),
```
Comment (yanndegat) [inline ovh/resource_cloud_project_kube_iprestrictions_test.go:59]: hi @IamBlueSlime 

sorry for the delay.

i think your test shall be : 
```
resource.TestCheckResourceAttr(
						"ovh_cloud_project_kube_iprestrictions.iprestrictions", "ips.0", "10.42.0.0/16"),
```

**Review by yanndegat — COMMENTED** [review]
hi @IamBlueSlime 

i think your test check on the list isnt written in the proper format

**Comment by LucasVanHaaren** [comment]
Hello, I'm very interested in this feature !

Maybe I can help you to release it ?

**Comment by rverchere** [comment]
Hi,

I'd love to see it merged in a future release, thanks! ;)

**Comment by scraly** [comment]
Hi @LucasVanHaaren and @IamBlueSlime,
Any news about it? 🙂

Thanks

**Comment by IamBlueSlime** [comment]
> Hi @LucasVanHaaren and @IamBlueSlime, Any news about it? 🙂
> 
> Thanks

I will not have the time to work on this anymore, I'll be happy if someone could manage to finish this work :)

**Comment by scraly** [comment]
This feature have been included in PR [274](https://github.com/ovh/terraform-provider-ovh/pull/274)

**Comment by scraly** [comment]
Thanks @IamBlueSlime for your work.
We included the feature in the PR #274.
I close this one 🙂

**Pull request closed without merging**